barretenberg/cpp/src/barretenberg/stdlib/primitives/group/straus_scalar_slice.cpp

```diff
@@ -30,82 +31,56 @@ straus_scalar_slice<Builder>::straus_scalar_slice(Builder* context,
                                                   const size_t table_bits)
     : _table_bits(table_bits)
 {
-    using FF = typename Builder::FF;
-    using witness_t = stdlib::witness_t<Builder>;
-
-    constexpr bool IS_ULTRA = Builder::CIRCUIT_TYPE == CircuitType::ULTRA;
-
     // convert an input cycle_scalar object into a vector of slices, each containing `table_bits` bits.
     // this also performs an implicit range check on the input slices
-    const auto slice_scalar = [&](const field_t& scalar, const size_t num_bits) {
+    const auto compute_scalar_slices =
+        [&](const field_t& scalar, const size_t num_bits) -> std::pair<std::vector<field_t>, std::vector<uint64_t>> {
+        const size_t num_slices = numeric::ceil_div(num_bits, table_bits);
+
         // we record the scalar slices both as field_t circuit elements and u64 values
         // (u64 values are used to index arrays and we don't want to repeatedly cast a stdlib value to a numeric
         // primitive as this gets expensive when repeated enough times)
-        std::pair<std::vector<field_t>, std::vector<uint64_t>> result;
-        result.first.reserve(static_cast<size_t>(1ULL) << table_bits);
-        result.second.reserve(static_cast<size_t>(1ULL) << table_bits);
+        std::vector<field_t> stdlib_slices;
+        std::vector<uint64_t> native_slices;
+        stdlib_slices.reserve(num_slices);
+        native_slices.reserve(num_slices);
 
         if (num_bits == 0) {
-            return result;
+            return { stdlib_slices, native_slices };
         }
+        // Case 1: If the scalar is constant, compute the slices natively
         if (scalar.is_constant()) {
-            const size_t num_slices = (num_bits + table_bits - 1) / table_bits;
             const uint64_t table_mask = (1ULL << table_bits) - 1ULL;
             uint256_t raw_value = scalar.get_value();
             for (size_t i = 0; i < num_slices; ++i) {
-                uint64_t slice_v = static_cast<uint64_t>(raw_value.data[0]) & table_mask;
-                result.first.push_back(field_t(slice_v));
-                result.second.push_back(slice_v);
+                uint64_t slice_value = static_cast<uint64_t>(raw_value.data[0]) & table_mask;
+                stdlib_slices.push_back(field_t(slice_value)); // constants equal to slice_value
+                native_slices.push_back(slice_value);
                 raw_value = raw_value >> table_bits;
             }
-
-            return result;
-        }
-        uint256_t raw_value = scalar.get_value();
```

Comment:
This loop was computing the native slices in the non-constant scalar case. This is redundant though because the next step is computing the stdlib slices via `decompose_into_default_range` which performs the exact same calculation, so I now simply extract the native slices via `get_value()`.

Reply:
nice!